### PR TITLE
HDDS-4122. Implement OM Delete Expired Open Key Request and Response

### PR DESCRIPTION
**Pull request by errose28**
## What changes were proposed in this pull request?

Implement OM request and response for moving keys from the open key table to the deleted table. These will be used as part of parent jira HDDS-4120 to implement the open key cleanup service. This pull request also moves some duplicate code between the existing OM key(s) delete request/response and the new open keys delete request/response to shared areas, and refactors the OM Key(s) delete request/response to use this shared code.

Volume and bucket level byte usage will eventually need to be updated by the new open keys delete request/response, but this is not done in this pull request. The current method used to do this in HDDS-541 can lead to DB inconsistency. When this is resolved in HDDS-4308, it can be added to the open key request/responses.

## What is the link to the Apache JIRA

HDDS-4122

## How was this patch tested?

Unit tests were added for the new OMOpenKeysDeleteRequest and OMOpenKeysDeleteResponse classes.

### CODE REVIEW

**Review by avijayanhwx — APPROVED** [review]
Thanks for working on this @errose28. Clean implementation on this. I am +1 on these changes as it is. Let's wait for a review from the OM team.

**Comment by bharatviswa504** [comment]
I will take a look at it today.
Thanks, @avijayanhwx for tagging.

**Review by bharatviswa504 — COMMENTED** [review]
Thank You @errose28 for the contribution and nice to see very extensive test coverage.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java:138]: Yes, the idea of creating this abstract class was to eventually consolidate the duplicate code between OMOpenKeysDeleteResponse, OMKeyDeleteResponse, and OMKeysDeleteResponse. I had originally refactored the other classes as well to use this code, but since HDDS-451 (quota support) is moving along at a brisk pace, I could not keep up with the merge conflicts as the other response classes kept changing, and decided it was better to do this in a later PR.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java:133]: Will fix.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java:118]: Just to clarify, is this the execution you are talking about?

1. Request1 deletes key1 from volume1 in cache.
2. Request2 deletes key2 from volume1 in cache.
3. Request1 sets cached VolumeArgs object volArgs.bytesUsed -= key1.bytesUsed.
    - *divergence 1*: The cache shows key1 and key2 as deleted, but cache byte usage only reflects key1's deletion.
4. Request2 sets cached VolumeArgs object volArgs.bytesUsed -= key2.bytesUsed.
    - At this point, byte usage in the cache is consistent with the keys it shows as deleted.
5. Response1 is processed, committing volArgs and the deletion of key1 to the DB.
    - *divergence 2*: the DB shows only key1 deleted, but volume byte usage has been set as if both key1 and key2 were deleted.
6. Response2 is processed, committing volArgs to the DB again, and committing the deletion of key2 to the DB.
    - Now the keys deleted and bytes used align in the DB.

IIRC the entire volume table is stored in memory and only persisted to the DB to save state. Reads only happen from the in memory cache for volume metadata. In this case, *divergence 2* will never be detected by callers since it only happens at the DB level. *divergence 1* my exist briefly and be detected by callers. Again, this is really an issue with all requests modified in HDDS-4053 and not just this PR. We should discuss to determine whether the slight inconsistency warrants a whole volume lock on all requests that modify byte usage.

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java:118]: Not exactly, what I mean here is by the time we process request1, as we add the same object to double buffer, and if other thread processing request2 and updating it will update DB state also (Technically this should happen after adding response to double buffer)

Cache is for holding in flight updates if it is not committed to DB, I see no issues with that, this is by design.

Diveregence 1 should not exist, if volume locks are held.

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java:118]: https://issues.apache.org/jira/browse/HDDS-2344 Jira but here just value updating (Might not be bringing ConcurrentModificationException, but it can provide some context)

Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java]: Thanks for the explanation @bharatviswa504. I now see that *divergence 2* in the above example poses an issue in the event of an OM crash happening between steps 5 and 6. This will cause the byte usage update to be applied twice in the DB after OM restart. Volume byte usage updates will be removed from the open key requests and responses. Since this is really a larger problem with all requests/responses operating in this way under HDDS-541, we can fix the issue when a solution is developed for all requests/responses.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/OMOpenKeysDeleteRequest.java]: Volume byte usage updates will be removed from the open keys delete request and response classes. See [this comment](https://github.com/apache/hadoop-ozone/pull/1435#discussion_r500389769).

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java:131]: Looks like development on the key(s) delete request and response classes has taken a break. I have refactored them to use these shared methods now.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java
Comment (bharatviswa504) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java:87]: Minor: Can we merge these two functions.

**Review by bharatviswa504 — COMMENTED** [review]
One minor comment.
Once that is addressed good to commit.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java
Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java:87]: This is actually related to a mistake I made in OMKeysDeleteResponse. The original implementation had one trxnLogIndex it used for all the keys. All other calls to this method are using the updateID of the keyInfo provided as the trxnLogIndex. If the way I am doing it currently (OMKeysDeleteResponse uses the updateID of each key as its trxnLogIndex instead of one value for all keys deleted), then I can remove the overload. If not, I can fix OMKeysDeleteResponse to call the overload, giving it identical behavior to its original implementation.

Comment (bharatviswa504) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java:87]: Previously we used to use updateID for detecting replay of transaction. Now we are not using updateID any more.
But from my understanding updateID should be set with transactionIndex even before.

Comment (errose28) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/response/key/AbstractOMKeyDeleteResponse.java:87]: Got it. A closer look at the OMKeysDeleteRequest looks like this was happening anyways. Instead of setting the update ID for the key info to be the trxnLogIndex and submitting the key info to the response, it was just passing the trxnLogIndex separately with the key info to the response. I will update OMKeysDeleteRequest/Response to be consistent with the other request/responses in how they do this, and remove the overload of this method.

**Comment by bharatviswa504** [comment]
Thank You @errose28 for the contribution and @avijayanhwx for the review

**Pull request merged**
